src/ophyd_async/tango/base_devices/base_device.py

```diff
@@ -0,0 +1,113 @@
+                    for variation in attribute names across similar servers.
+    """
+
+    src_dict: dict = {}
+    trl: str = ""
+    proxy: Optional[DeviceProxy] = None
+
+    # --------------------------------------------------------------------
+    def __init__(self, trl: str, name="") -> None:
+        self.trl = trl
+        self.proxy: Optional[DeviceProxy] = None
+        self.create_children_from_annotations()
+        StandardReadable.__init__(self, name=name)
+
+    async def connect(
+        self,
+        mock: bool = False,
+        timeout: float = DEFAULT_TIMEOUT,
+        force_reconnect: bool = False,
+    ):
+        async def closure():
+            if self.proxy is None:
+                self.proxy = await DeviceProxy(self.trl)
+            elif isinstance(self.proxy, asyncio.Future):
+                self.proxy = await self.proxy
+            return self
+
+        await closure()
+        self.register_signals()
+        # set_name should be called again to propagate the new signal names
+        self.set_name(self.name)
+
+        await super().connect(mock=mock, timeout=timeout)
+
+    def register_signals(self):
+        for name, obj_type in self.__annotations__.items():
+            if hasattr(self, name):
+                signal = getattr(self, name)
+                if issubclass(type(signal), Signal):
+                    tango_name = name.lstrip("_")
+                    read_trl = f"{self.trl}/{tango_name}"
+                    datatype = infer_python_type(read_trl)
+                    backend = make_backend(
+                        datatype=datatype,
+                        read_trl=read_trl,
+                        write_trl=read_trl,
+                        device_proxy=self.proxy,
+                    )
+                    signal._backend = backend  # noqa: SLF001
+
+    def create_children_from_annotations(self):
+        for attr_name, obj_type in self.__annotations__.items():
+            if (
+                isinstance(obj_type, type)
+                and issubclass(obj_type, Signal)
+                or obj_type is None
+            ):
+                if obj_type is SignalRW:
+                    setattr(self, attr_name, SignalRW())
+                elif obj_type is SignalR:
+                    setattr(self, attr_name, SignalR())
+                elif obj_type is SignalW:
+                    setattr(self, attr_name, SignalW())
+                elif obj_type is SignalX:
+                    setattr(self, attr_name, SignalX())
+                elif obj_type is Signal or obj_type is None:
+                    tango_name = attr_name.lstrip("_")
+                    setattr(
+                        self,
+                        attr_name,
+                        infer_signal_frontend(trl=f"{self.trl}/" f"{tango_name}"),
+                    )
+                else:
+                    print(obj_type, type(obj_type))
+                    raise ValueError(f"Invalid signal type {obj_type}")
```

Comment:
Could you use https://github.com/bluesky/ophyd-async/blob/d7f07488fdfa976037791cbdfdb21e0f7c5f7430/src/ophyd_async/epics/pvi/_pvi.py#L309-L341 if we put it in `ophyd_async.core`?

Reply:
@coretl this doesn't seem to work at the moment for StandardReadables. This throws an error for _readables because it is a Tuple.

In principle yes, we could use this but it does not allow for inferring the signal frontend and datatype from tango. This is fine but it means we wouldn't be taking advantage of this Tango feature. My instinct would be to overwrite this with a tango version. Perhaps it could accept a callback method to run if you want the frontend to be inferred.

```python
  File "/home/burkedev/projects/BLUESKY/ophyd-async/src/ophyd_async/tango/base_devices/base_device.py", line 50, in __init__
    create_children_from_annotations(self)
  File "/home/burkedev/projects/BLUESKY/ophyd-async/src/ophyd_async/epics/pvi/_pvi.py", line 342, in create_children_from_annotations
    sub_device = device_type()
                 ^^^^^^^^^^^^^
  File "/usr/lib/python3.11/typing.py", line 1246, in __call__
    raise TypeError(f"Type {self._name} cannot be instantiated; "
TypeError: Type Tuple cannot be instantiated; use tuple() instead
```

Reply:
> @coretl this doesn't seem to work at the moment for StandardReadables. This throws an error for _readables because it is a Tuple.

That should be relatively easy to fix, we should do that

> In principle yes, we could use this but it does not allow for inferring the signal frontend and datatype from tango. This is fine but it means we wouldn't be taking advantage of this Tango feature. My instinct would be to overwrite this with a tango version. Perhaps it could accept a callback method to run if you want the frontend to be inferred.

I'm not sure I understand what the signal frontend is?

As for the datatype, we want to be able to support both specified signals that have specified types, and other unspecified signals that take their type from the control system. For example, imagine a PVI Device that has a `specified_sig` that gives you a `float` when you `caget` it, and an `unspecified_sig` that gives you a `str` when you `caget` it. Then you can:

```python
class MyDevice(PviDevice):
   specified_sig: SignalRW[float]

with DeviceCollector():
    device = MyDevice("PV:PREFIX")
# Now device has both specified_sig and unspecified_sig

await device.specified_sig.put(32.0)  # passes at static analysis and runtime
await device.specified_sig.put("bad")  # fails at static analysis and runtime

RE(save_device(device))  # gets specified_sig and unspecified_sig values
await device.unspecified_sig.put(32.0)  # fails at static analysis but passes at runtime
```

This means that:
- Signals we specify the type of will have their usage checked at static analysis time, then at runtime will be check to ensure the control system type matches
- Signals we don't specify will be added to the device using the control system type. This allows generic load/save to include them, and at runtime you can use the attribute but static analysis will fail if you use it in code
- If you start using an unspecified signal in code, then fix its type and make it a specified signal

Reply:
I write tango-specific solutions to create children from annotations and to fill signals from the tango device server. This will be adapted to a system agnostic solution once one has been finalized.